src/main/java/com/pusher/client/channel/impl/PrivateEncryptedChannelImpl.java

```diff
@@ -124,6 +127,72 @@ public void updateState(ChannelState state) {
         }
     }
 
+    @Override
+    public PusherEvent prepareEvent(String event, String message) {
+        try {
+
+            Map receivedMessage = GSON.fromJson(message, Map.class);
+            final String decryptedMessage = decryptMessage((String) receivedMessage.get("data"));
+            receivedMessage.replace("data", decryptedMessage);
+
+            return GSON.fromJson(
+                    GSON.toJson(receivedMessage), PusherEvent.class);
+
+        } catch (AuthenticityException e1) {
+
+            // retry once only.
+            disposeSecretBoxOpener();
+            authenticate();
+
+            try {
+                Map receivedMessage = GSON.fromJson(message, Map.class);
+                final String decryptedMessage = decryptMessage((String) receivedMessage.get("data"));
+                receivedMessage.replace("data", decryptedMessage);
+
+                return GSON.fromJson(
+                        GSON.toJson(receivedMessage), PusherEvent.class);
+            } catch (AuthenticityException e2) {
+                disposeSecretBoxOpener();
+                notifyListenersOfDecryptFailure(event);
+            }
+        }
```

Comment:
i'm not sure I like this nested try/catch but I'm not sure how to improve it... 🤔

Reply:
Yes, it's not ideal, although it kind of works for a single retry. If we wanted multiple retries, we might prefer a loop:

```
for (int attempt = 0; attempt < MAX_ATTEMPTS; attempt++) {
  try {
    // decrypt it
    return decryptedMessage;
  } catch {
    // dispose of the secret
    // authenticate (possibly unless attempt == MAX_ATTEMPTS-1)
  }
}

// notify listeners of failure because we ran out of attempts before returning
return null;
```

Alternatively, you can un-nest the second attempt by recognising that the `try` block ends with `return`, so the only way to reach code after the whole `try`/`catch` construct is via the `catch`.

```
try {
  // decrypt it
  return decryptedMessage;
} catch {
  // dispose of the secret
}

// authenticate

try {
  // decrypt it
  return decryptedMessage;
} catch {
  // dispose of the secret
  // notify listeners
}
```

It might also benefit from a private function to encapsulated the repeated code in the decryption attempt.

None of them seem super elegant, but some food for thought.

Reply:
Thanks for the helpful response. I'll have a think over this and see what I can do to refactor it to be a little bit better :)

Reply:
I've refactored out Json parts to decryptMessage so it is now responsible for returning the PusherEvent. I moved the retry logic into its own method so hopefully it's easier to follow what's happening instead of nested try's and catches.